replica/table.cc

```diff
@@ -550,17 +550,28 @@ void table::enable_off_strategy_trigger() {
 
 storage_group_manager::~storage_group_manager() = default;
 
-future<> storage_group_manager::parallel_foreach_storage_group(std::function<future<>(size_t, storage_group&)> f) {
-    rwlock::holder shared_lock = co_await get_rwlock().hold_read_lock();
-    co_await coroutine::parallel_for_each(_storage_groups, [&] (std::pair<size_t, const std::unique_ptr<storage_group>&> p) -> future<> {
-        co_await f(p.first, *p.second.get());
+// exception-less attempt to hold gate.
+static std::optional<gate::holder> try_hold_gate(gate& g) noexcept {
+    return g.is_closed() ? std::nullopt : std::make_optional(g.hold());
+}
+
+future<> storage_group_manager::parallel_foreach_storage_group(std::function<future<>(storage_group&)> f) {
+    auto storage_groups = boost::copy_range<std::vector<storage_group_ptr>>(_storage_groups | boost::adaptors::map_values);
+    co_await coroutine::parallel_for_each(storage_groups, [&] (const storage_group_ptr& sg) -> future<> {
```

Comment:
parallel_for_each doesn't yield when iterating the range so the copy above is superfluous.
The lambda can accept a `storage_group_ptr` by value instead to extend its lifetime (besides entering its gate)

Reply:
I think we discussed this in previous patch and was ok with this extra copy here to make it explicit the intent, rather than relying on implementation detail of parallel_for_each(). table wide ops aren't part of hot path. do you agree?

Reply:
You're right, it's fine as long as we make it an invariant.